src/triage/component/catwalk/utils.py

```diff
@@ -190,14 +190,21 @@ def save_db_objects(db_engine, db_objects):
 
     Args:
         db_engine (sqlalchemy.engine)
-        db_objects (list) SQLAlchemy model objects, corresponding to a valid table
+        db_objects (iterable) SQLAlchemy model objects, corresponding to a valid table
     """
     with tempfile.TemporaryFile(mode='w+') as f:
```

Comment:
To be honest, maybe optimizing for low memory usage by using disk is not right here. Maybe we should assume this is on a big machine (as it usually is for us) and optimize for low disk I/O by using a StringIO/BytesIO?

Reply:
It's generally pretty straightforward to write CSV to standard out and ask PostgreSQL to read from standard in, no? That way you optimize both and have no scaling issue.

`postgres_copy` is in fact executing `COPY … FROM STDIN`, but accepts any file object to pipe to PostgreSQL. I imagine you might be able to use standard out; but, since what you have direct access to is all in-process, perhaps it would be easiest to hand `postgres_copy` your own file object which streams the data.

Of course it would be great if `postgres_copy` just wanted an iterable. The Python file object provides the same interface. But, it looks like in this case `psycopg` really wants something from `io`.

This is something I've come across before, and which seems to be a perennial issue here; so, I'm happy to play around with it, in hope I can help.

It doesn't seem as though Python offers a ready-built class for streaming text from an iterable, but this might work –

    class IterStream(io.RawIOBase):

        def __init__(self, iterable):
            self.leftover = None
            self.iterable = iter(iterable)

        def readable(self):
            return True

        def readinto(self, b):
            try:
                chunk = self.leftover or next(self.iterable)
            except StopIteration:
                return 0  # indicate EOF

            l = len(b)  # We're supposed to return at most this much
            (output, self.leftover) = (chunk[:l], chunk[l:])
            b[:len(output)] = output
            return len(output)

    def open_iter(iterable, buffer_size=io.DEFAULT_BUFFER_SIZE):
        return io.BufferedReader(IterStream(iterable), buffer_size=buffer_size)

[[Stackoverflow]](https://stackoverflow.com/questions/6657820/python-convert-an-iterable-to-a-stream/20260030#20260030)

Of course, here, we have a special case. It seems as though we could improve the above example, since `csv` wants to write to a file object; but, for the moment, let's just connect it, if hackily –

    def iter_csv(iterable, *args, **kwargs):
        output = io.StringIO()
        writer = csv.writer(output, *args, **kwargs)

        for row in iterable:
            output.seek(0)
            output.truncate()

            writer.writerow(row)
            yield output.getvalue()

Then, we could just use a generator expression one-liner here, but … let's not –

    def model_to_row(db_objects):
        for db_object in db_objects:
            yield [getattr(db_object, col.name)
                   for col in db_object.__table__.columns]

And so finally:

    def save_db_objects(db_engine, db_objects):
        db_objects = iter(db_objects)

        try:
            first_object = next(db_objects)
        except StopIteration:
            # nothing to do, why continue?
            return
        else:
            # restore full stream
            db_objects = itertools.chain((first_object,), db_objects)

        rows = model_to_row(db_objects)
        csv = iter_csv(rows, quoting=csv.QUOTE_MINIMAL)
        with open_iter(csv) as iter_descriptor:
            postgres_copy.copy_from(iter_descriptor, type(first_object), db_engine, format='csv')

Reply:
@ecsalomon Thank you for reviving this PR -- I was just looking for this comment, and I couldn't find it!!